spec/services/hmpps_api/assessment_api_spec.rb

```diff
@@ -0,0 +1,75 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe HmppsApi::AssessmentApi do
+  let(:api_host) { Rails.configuration.assessment_api_host }
+
+  before do
+    stub_auth_token
+  end
+
+  describe '.get_latest_oasys_date' do
+    let(:offender_no) { 'A9346AC' }
+
+    let(:stub_url) { "#{api_host}/offenders/nomisId/#{offender_no}/assessments/summary?assessmentType=LAYER_3&assessmentStatus=COMPLETE" }
+
+    context 'when there is one completed OASys assessment' do
+      # The last assessment date is the 'completed' field.
+      # The test expects it to be a date object for consistency throughout the code base
+      let(:completed_date) { '2012-12-10'.to_date }
+      let(:latest_assessment) { build(:assessment_api_response, completed: completed_date) }
+
+      before do
+        stub_request(:get, stub_url).to_return(status: 200, body: [
+          latest_assessment].to_json)
+      end
+
+      it 'checks that the API has only been called once' do
+        described_class.get_latest_oasys_date(offender_no)
+        expect(a_request(:get, stub_url)).to have_been_made.once
+      end
+
+      it 'returns the date the assessment was completed' do
+        expect(described_class.get_latest_oasys_date(offender_no)).to eq(completed_date)
+      end
+    end
+
+    context 'when there are multiple completed OASys assessments' do
+      let(:old_assessments) { build_list(:assessment_api_response, 9) }
+      let(:latest_assessment) { build(:assessment_api_response, completed: Time.zone.now) }
+      let(:all_assessments) { (old_assessments + [latest_assessment]).shuffle }
+
+      before do
+        stub_request(:get, stub_url).to_return(status: 200, body: all_assessments.to_json)
+      end
+
+      it 'returns the most recently completed assessment' do
+        expect(described_class.get_latest_oasys_date(offender_no)).to eq(Time.zone.today)
+      end
+    end
+
+    context 'when there are no assessments (it returns an empty array)' do
+      before do
+        stub_request(:get, stub_url).to_return(status: 200, body: [].to_json)
+      end
+
+      it 'returns nil' do
+        expect(described_class.get_latest_oasys_date(offender_no)).to eq(nil)
+      end
+    end
+
+    context 'when the offender is not found (status 404)' do
+      before do
+        stub_request(:get, stub_url).to_return(status: 200, body: {
+          "status": 404,
+          "developerMessage": "Offender not found for NOMIS, A1857ER"
+        }.to_json)
```

Comment:
It looks like the server actually response with a `404 Not Found` status code in this scenario, not a `200 OK`.

I know we talked about it on Friday and said that it was weird considering the response body suggests we should be receiving a `404` error. So it's actually a relief to see that it actually does the right thing and returns a sensible error code.

You can see the response status code by adding the option `-i` to your `curl` command. For example, with the non-existent offender ID `A1234ZC`:

```
$ curl -i "https://offender-dev.aks-dev-1.studio-hosting.service.justice.gov.uk/offenders/nomisId/A1234ZC/assessments/summary?assessmentType=LAYER_3&assessmentStatus=COMPLETE" --header "Content-Type: application/json" --header "Authorization: Bearer $TOKEN"
HTTP/2 404
server: nginx/1.19.2
date: Mon, 27 Sep 2021 16:58:40 GMT
content-type: application/json
vary: Accept-Encoding
request-context: appId=0d5d6f09-1d07-47f5-9d3a-f9f473544e78
x-content-type-options: nosniff
x-xss-protection: 1; mode=block
cache-control: no-cache, no-store, max-age=0, must-revalidate
pragma: no-cache
expires: 0
strict-transport-security: max-age=15724800; includeSubDomains
x-frame-options: SAMEORIGIN

{"status":404,"developerMessage":"Offender not found for NOMIS, A1234ZC"}
```

Note the first line to be returned from `curl` is:

```
HTTP/2 404
```

---

We should change the stubbed request accordingly so that it returns a `404` response.

And then we want our API client to handle that error response gracefully. I think the described behaviour of "it returns nil" is what we want.